happytransformer/trainers/trainer_qa.py

```diff
@@ -0,0 +1,192 @@
+
+        result_output = "Evaluating Result: " + str(correct) + "/" + str(total) + " -- " + ending
+        self.logger.info(result_output)
+
+        if output_filepath != None:
+            fieldnames = ["contexts", "questions", "answer", "outputs", "correct"]
+            self._output_result_to_csv(output_filepath, fieldnames, results)
+
+        return score
+
+
+    def test(self, filepath, args, output_filepath):
+        #todo
+        pass
+
+    def __get_train_eval_data(self, filepath):
+        """
+        Used for parsing data for training and evaluating (both contain labels)
+        :param filepath: a string that contains the location of the data
+        :return:
+        """
+
+        contexts = []
+        questions = []
+        answers = []
+        with open(filepath, newline='') as csv_file:
+            reader = csv.DictReader(csv_file)
+            for row in reader:
+                contexts.append(row['context'])
+                questions.append(row['question'])
+                answer = {}
+                answer["answer_text"] = row['answer_text']
+                answer["answer_start"] = int(row['answer_start'])
+                answers.append(answer)
+        csv_file.close()
+
+        return contexts, questions, answers
+
+
+    def __add_end_idx(self, contexts, answers):
+        for answer, context in zip(answers, contexts):
+
+            gold_text = answer['answer_text']
+            start_idx = answer['answer_start']
+            end_idx = start_idx + len(gold_text)
+
+            # todo (maybe): strip answer['text'] (remove white space from start and end)
+            # sometimes squad answers are off by a character or two – fix this
+            if context[start_idx:end_idx] == gold_text:
+                answer['answer_end'] = end_idx
+            elif context[start_idx - 1:end_idx - 1] == gold_text:
+                answer['answer_start'] = start_idx - 1
+                answer['answer_end'] = end_idx - 1  # When the gold label is off by one character
+            elif context[start_idx - 2:end_idx - 2] == gold_text:
+                answer['answer_start'] = start_idx - 2
+                answer['answer_end'] = end_idx - 2  # When the gold label is off by two characters
+            else:
+                print("error: implement skipping training answer")
+
+    def __add_token_positions(self, encodings, answers):
+        start_positions = []
+        end_positions = []
+        for i in range(len(answers)):
+            start_positions.append(encodings.char_to_token(i, answers[i]['answer_start']))
+            end_positions.append(encodings.char_to_token(i, answers[i]['answer_end'] - 1))
+            # if None, the answer passage has been truncated
+            if start_positions[-1] is None:
+                start_positions[-1] = self.tokenizer.model_max_length
+            if end_positions[-1] is None:
+                end_positions[-1] = self.tokenizer.model_max_length
```

Comment:
Retroactively fixing up the positions seems odd.

Reply:
I agree this seems a little odd. I followed the official documentation. In the future we can adjust it so hopefully we don't have to retroactively fix the positions. But since it's within the official documentation, it likely works.